opm/simulators/linalg/bda/amgclSolverBackend.cpp

```diff
@@ -199,9 +199,39 @@ void amgclSolverBackend<block_size>::convert_data(double *vals, int *rows) {
 } // end convert_data()
 
 #if HAVE_VEXCL
+void initialize_vexcl(std::unique_ptr<vex::Context>& ctx, unsigned int platformID, unsigned int deviceID) {
+    std::vector<cl::Platform> platforms;
+    cl::Platform::get(&platforms);
+
+    if (platforms.size() <= platformID) {
+        OPM_THROW(std::logic_error, "Error chosen too high OpenCL platform ID");
+    }
+
+    std::string opencl_info;
+    platforms[platformID].getInfo(CL_PLATFORM_NAME, &opencl_info);
+
+    setenv("OCL_PLATFORM", opencl_info.c_str(), true); // overwrites if OCL_PLATFORM is already set
+
+    std::vector<cl::Device> devices;
+    platforms[platformID].getDevices(CL_DEVICE_TYPE_ALL, &devices);
+
+    if (devices.size() <= deviceID){
+        OPM_THROW(std::logic_error, "Error chosen too high OpenCL device ID");
+    }
+
+    devices[deviceID].getInfo(CL_DEVICE_NAME, &opencl_info);
+    setenv("OCL_DEVICE", opencl_info.c_str(), true); // overwrites if OCL_DEVICE is already set
```

Comment:
Using environment filter when you already have a `cl::Device` reference is a bit roundabout. `vex::Context` in vexcl is just a convenience wrapper, and all it is needed for is filter functionality for easy initialization. It is always implicitly converted to `std::vector<cl::CommandQueue>` for such things as `vex::vector` constructor, so I would use `std::vector<cl:CommandQueue>` directly instead of `vex::Context` here. All you need to do at this point is something like
```cpp
cl::Context c(devices[deviceID]);
cl::CommandQueue q(c, devices[deviceID]);
ctx.push_back(q); // ctx is now std::vector<cl::CommandQueue>

```

Reply:
I've overwritten it with 796d1b7